File: app/src/main/java/dev/ragnarok/fenrir/model/SideDrawerCategory.java
package dev.ragnarok.fenrir.model;

import android.os.Parcel;
import android.os.Parcelable;

import androidx.annotation.StringRes;

public class SideDrawerCategory implements Parcelable {

    public static Creator<SideDrawerCategory> CREATOR = new Creator<SideDrawerCategory>() {
        public SideDrawerCategory createFromParcel(Parcel source) {
            return new SideDrawerCategory(source);
        }

        public SideDrawerCategory[] newArray(int size) {
            return new SideDrawerCategory[size];
        }
    };
    @StringRes
    private final int title;
    @SideSwitchableCategory
    private final int key;
    private boolean checked;

    public SideDrawerCategory(@SwitchableCategory int key, @StringRes int title) {
        this.title = title;
        this.key = key;
    }

    public SideDrawerCategory(Parcel in) {
        title = in.readInt();
        //noinspection ResourceType
        key = in.readInt();
        checked = in.readInt() == 1;
    }

    @StringRes
    public int getTitle() {
        return title;
    }

    @SideSwitchableCategory
    public int getKey() {
        return key;
    }

    public boolean isChecked() {
        return checked;
    }

    public SideDrawerCategory setChecked(boolean checked) {
        this.checked = checked;
        return this;
    }

    @Override
    public int describeContents() {
        return 0;
    }

    @Override
    public void writeToParcel(Parcel dest, int flags) {
        dest.writeInt(title);
        dest.writeInt(key);
        dest.writeInt(checked ? 1 : 0);
    }
}

File: app_fenrir/src/main/kotlin/dev/ragnarok/fenrir/model/SideDrawerCategory.kt
package dev.ragnarok.fenrir.model

import android.os.Parcel
import android.os.Parcelable
import androidx.annotation.StringRes
import dev.ragnarok.fenrir.getBoolean
import dev.ragnarok.fenrir.putBoolean

class SideDrawerCategory : Parcelable {
    @StringRes
    private val title: Int

    @SideSwitchableCategory
    private val key: Int
    private var checked = false

    constructor(@SwitchableCategory key: Int, @StringRes title: Int) {
        this.title = title
        this.key = key
    }

    internal constructor(`in`: Parcel) {
        title = `in`.readInt()
        key = `in`.readInt()
        checked = `in`.getBoolean()
    }

    @StringRes
    fun getTitle(): Int {
        return title
    }

    @SideSwitchableCategory
    fun getKey(): Int {
        return key
    }

    fun isChecked(): Boolean {
        return checked
    }

    fun setChecked(checked: Boolean): SideDrawerCategory {
        this.checked = checked
        return this
    }

    override fun describeContents(): Int {
        return 0
    }

    override fun writeToParcel(dest: Parcel, flags: Int) {
        dest.writeInt(title)
        dest.writeInt(key)
        dest.putBoolean(checked)
    }

    companion object CREATOR : Parcelable.Creator<SideDrawerCategory> {
        override fun createFromParcel(parcel: Parcel): SideDrawerCategory {
            return SideDrawerCategory(parcel)
        }

        override fun newArray(size: Int): Array<SideDrawerCategory?> {
            return arrayOfNulls(size)
        }
    }
}